### ISSUE TITLE
NullPointerException Explanations after trying to delete Constraint

### ISSUE THREAD

**Issue by Kogoro** [FeatureIDE/FeatureIDE#512]
#### Prerequisitives
* **Eclipse Version:**  `Neon.3`
* **FeatureIDE-Version:**  `3.3.0`

### Issue description
Getting this nullpointer when trying to delete a constraint that is contained in an explanation.
```
java.lang.IndexOutOfBoundsException: Index: 3, Size: 3
	at java.util.ArrayList.rangeCheck(Unknown Source)
	at java.util.ArrayList.get(Unknown Source)
	at java.util.Collections$UnmodifiableList.get(Unknown Source)
	at de.ovgu.featureide.fm.ui.editors.elements.GraphicalFeatureModel.getConstraintIndex(GraphicalFeatureModel.java:238)
	at de.ovgu.featureide.fm.ui.editors.featuremodel.figures.LegendFigure.createExplanationEntry(LegendFigure.java:724)
	at de.ovgu.featureide.fm.ui.editors.featuremodel.figures.LegendFigure.createRows(LegendFigure.java:384)
	at de.ovgu.featureide.fm.ui.editors.featuremodel.figures.LegendFigure.<init>(LegendFigure.java:178)
	at de.ovgu.featureide.fm.ui.editors.featuremodel.editparts.LegendEditPart.createFigure(LegendEditPart.java:58)
	at de.ovgu.featureide.fm.ui.editors.featuremodel.editparts.LegendEditPart.createFigure(LegendEditPart.java:1)
	at org.eclipse.gef.editparts.AbstractGraphicalEditPart.getFigure(AbstractGraphicalEditPart.java:494)
	at de.ovgu.featureide.fm.ui.editors.featuremodel.editparts.LegendEditPart.getFigure(LegendEditPart.java:53)
	at de.ovgu.featureide.fm.ui.editors.featuremodel.editparts.LegendEditPart.getFigure(LegendEditPart.java:1)
	at org.eclipse.gef.editparts.AbstractGraphicalEditPart.isSelectable(AbstractGraphicalEditPart.java:567)
	at org.eclipse.gef.editparts.AbstractEditPart.getTargetEditPart(AbstractEditPart.java:632)
	at org.eclipse.gef.tools.MarqueeSelectionTool.isMarqueeSelectable(MarqueeSelectionTool.java:559)
	at org.eclipse.gef.tools.MarqueeSelectionTool.calculatePrimaryMarqueeSelectedEditParts(MarqueeSelectionTool.java:323)
	at org.eclipse.gef.tools.MarqueeSelectionTool.calculateMarqueeSelectedEditParts(MarqueeSelectionTool.java:278)
	at org.eclipse.gef.tools.MarqueeSelectionTool.performMarqueeSelect(MarqueeSelectionTool.java:722)
	at org.eclipse.gef.tools.MarqueeSelectionTool.handleButtonUp(MarqueeSelectionTool.java:487)
	at org.eclipse.gef.tools.AbstractTool.mouseUp(AbstractTool.java:1200)
	at org.eclipse.gef.tools.SelectionTool.mouseUp(SelectionTool.java:574)
	at org.eclipse.gef.EditDomain.mouseUp(EditDomain.java:301)
	at org.eclipse.gef.ui.parts.DomainEventDispatcher.dispatchMouseReleased(DomainEventDispatcher.java:380)
	at org.eclipse.draw2d.LightweightSystem$EventHandler.mouseUp(LightweightSystem.java:548)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:221)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:84)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4418)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1079)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4236)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3824)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$4.run(PartRenderingEngine.java:1121)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1022)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:150)
	at org.eclipse.ui.internal.Workbench$5.run(Workbench.java:693)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:610)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:148)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:138)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:388)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:243)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:673)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:610)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1519)
```

**Comment by tthuem** [comment]
@TimoGuenther Please try to fix that easter egg.

**Comment by TimoGuenther** [comment]
When a redundant constraint is selected, the legend includes the index of the constraint, obtained via `GraphicalFeatureModel#getConstraintIndex(Constraint)`. Unfortunately, that method uses the size of `this.constraints` (which is a map) to determine how many constraints there are, but then actually iterates over `getFeatureModel().getConstraints()` (because it is a list, whereas maps are unordered). Of course, this operates under the assumption that both of these collections have the same size. However, after deleting a constraint, the program is in a limbo state where this is not true, resulting in the `IndexOutOfBoundsException` above.

This means there are several possible solutions to this issue:
1. Make `GraphicalFeatureModel#getConstraintIndex(Constraint)` use the same collection for size detection and iteration. This should probably be the case anyway.
2. Make sure both collections always have the same size. This is possibly a non-trivial change.
3. Do not display the index of the constraint in the legend so `GraphicalFeatureModel#getConstraintIndex(Constraint)` does not have to be called. The legend could just say "the selected constraint" and be done with it.

**Comment by tthuem** [comment]
I would say for now, just omit the number of the constraint. The rest sounds reasonable for the next release.

**Comment by TimoGuenther** [comment]
Implemented approach 3 in 4f2ab55a0644c48c63c993d483901f47286b2095.
